File: examples/cpp/shapes.cpp
// shapes.cpp — geometric shapes used to exercise the C++ language capabilities.
// Deterministic: no I/O, no randomness, no timestamps.

#include <string>

// PI is a compile-time constant used by the area helpers.
const double PI = 3.14159265358979323846;

// area computes the area of a circle from its radius. This FREE function shares
// its name with Circle::area to exercise SymbolID disambiguation by container.
double area(double radius) {
    return PI * radius * radius;
}

/// Circle is a simple shape with a radius.
/// It demonstrates a documented class with member functions.
class Circle {
public:
    // Circle constructs a circle of the given radius.
    Circle(double r) : radius_(r) {}

    // area returns the area of this circle. Same name as the free function
    // area(double) above, but reached via container "Circle".
    double area() const {
        return PI * radius_ * radius_;
    }

    // circumference returns the perimeter of the circle.
    double circumference() const {
        return 2.0 * PI * radius_;
    }

private:
    double radius_;
};

/* Rectangle is a width-by-height shape.
   This block comment documents the struct. */
struct Rectangle {
    double width;
    double height;

    // area returns width times height.
    double area() const {
        return width * height;
    }
};

// describe returns a human-readable label for a shape name. It has a multi-line
// doc comment so doc-extraction can be checked against an exact expected string.
std::string describe(const std::string& kind) {
    return "shape: " + kind;
}

File: examples/python/shapes.py
"""Geometry helpers used to exercise the Python language reader/writer.

This module mixes documented and undocumented module-level functions, classes
with methods (to exercise SymbolID.Container), and names that are reused across
containers (to exercise disambiguation).
"""


def area(width, height):
    """Return the area of a rectangle."""
    return width * height


def describe(name):
    return "shape: " + name


def perimeter(width, height):
    """Return the rectangle perimeter.

    The perimeter is twice the sum of the two sides.
    """
    return 2 * (width + height)


class Rectangle:
    """A rectangle defined by its width and height."""

    def __init__(self, width, height):
        self.width = width
        self.height = height

    def area(self):
        """Return the rectangle area."""
        return self.width * self.height

    def describe(self):
        return "rectangle"


class Circle:
    """A circle defined by its radius."""

    def __init__(self, radius):
        self.radius = radius

    def area(self):
        return 3 * self.radius * self.radius
